### PR TITLE
Transaction issues for alert reminder

### LINKED ISSUE #13389: Grafana 5.3.0-beta1 /login hang
After upgrading Grafana from 5.2.4 to 5.3.0-beta1, requests to load `/login` started to hang and eventually time out.

If I add `[alerting]` `execute_alerts = false` in the config, everything appears to work normally and the instance loads quickly.

I do not see excessive CPU or memory usage, and I do not see any long-running queries in MySQL.

```
# time curl -vvv --max-time 10 http://localhost:3000/login
*   Trying ::1...
* TCP_NODELAY set
* Connected to localhost (::1) port 3000 (#0)
> GET /login HTTP/1.1
> Host: localhost:3000
> User-Agent: curl/7.52.1
> Accept: */*
>
* Operation timed out after 10001 milliseconds with 0 bytes received
* Curl_http_done: called premature == 1
* stopped the pause stream!
* Closing connection 0
curl: (28) Operation timed out after 10001 milliseconds with 0 bytes received

real    0m10.014s
user    0m0.004s
sys     0m0.008s
```

requests to / return immediately:

```
# time curl -vvv --max-time 10 http://localhost:3000/
*   Trying ::1...
* TCP_NODELAY set
* Connected to localhost (::1) port 3000 (#0)
> GET / HTTP/1.1
> Host: localhost:3000
> User-Agent: curl/7.52.1
> Accept: */*
>
< HTTP/1.1 302 Found
< Content-Type: text/html; charset=utf-8
< Location: /login
< Set-Cookie: grafana_sess=3165e067866bb724; Path=/; HttpOnly
< Set-Cookie: redirect_to=%252F; Path=/; HttpOnly
< Date: Fri, 21 Sep 2018 21:42:26 GMT
< Content-Length: 29
<
<a href="/login">Found</a>.

* Curl_http_done: called premature == 0
* Connection #0 to host localhost left intact

real    0m0.069s
user    0m0.006s
sys     0m0.004s
```

### LINKED ISSUE #13492: Alert reminders deduping not working as expected.

### PR DESCRIPTION
**Pull request by bergquist**
- [x] rename `alert_notification_journaling` to `alert_notification_state`
- [x] implement sql queries
- [x] implement notification steps in notifier.go
- [x] Test in HA mode with marefr new ha testing scripts

Fixes #13389 
Closes #13492

### CODE REVIEW

File: pkg/models/alert_notifications.go [hunk 1]
```diff
@@ -76,33 +85,30 @@ type GetAllAlertNotificationsQuery struct {
 	Result []*AlertNotification
 }
 
-type AlertNotificationJournal struct {
+type AlertNotificationState struct {
 	Id         int64
 	OrgId      int64
 	AlertId    int64
 	NotifierId int64
 	SentAt     int64
```
Comment (bergquist) [inline pkg/models/alert_notifications.go]: Remove `SentAt` and use `UpdatedAt` for all timestamping neeeeds.

File: pkg/models/alert_notifications.go [hunk 2]
```diff
@@ -76,33 +85,30 @@ type GetAllAlertNotificationsQuery struct {
...
 
-type RecordNotificationJournalCommand struct {
-	OrgId      int64
-	AlertId    int64
-	NotifierId int64
-	SentAt     int64
-	Success    bool
+type SetAlertNotificationStateToPendingCommand struct {
+	AlertRuleStateUpdatedVersion int64
+	State                        *AlertNotificationState
```
Comment (bergquist) [inline pkg/models/alert_notifications.go]: Don't pass state. pass each property and return state.

File: pkg/models/alert_notifications.go [hunk 3]
```diff
@@ -76,33 +85,30 @@ type GetAllAlertNotificationsQuery struct {
...
-	AlertId    int64
-	NotifierId int64
-
-	Result []AlertNotificationJournal
+type SetAlertNotificationStateToCompleteCommand struct {
+	State *AlertNotificationState
 }
 
-type CleanNotificationJournalCommand struct {
+type GetNotificationStateQuery struct {
```
Comment (bergquist) [inline pkg/models/alert_notifications.go]: GetOrCreateNotificationStateQuery

File: pkg/services/alerting/interfaces.go [hunk 4]
```diff
@@ -20,19 +22,24 @@ type Notifier interface {
...
+	ShouldNotify(ctx context.Context, evalContext *EvalContext, notificationState *models.AlertNotificationState) bool
 
 	GetNotifierId() int64
 	GetIsDefault() bool
 	GetSendReminder() bool
 	GetFrequency() time.Duration
 }
 
-type NotifierSlice []Notifier
+type NotifierState struct {
```
Comment (bergquist) [inline pkg/services/alerting/interfaces.go]: can we avoid exporting this struct?

File: pkg/services/alerting/notifier.go [hunk 5]
```diff
@@ -41,61 +40,82 @@ type notificationService struct {
...
 
-	return n.sendNotifications(context, notifiers)
+	return n.sendNotifications(context, notifierStates)
 }
 
-func (n *notificationService) sendNotifications(evalContext *EvalContext, notifiers []Notifier) error {
-	for _, notifier := range notifiers {
-		not := notifier
+func (n *notificationService) sendAndMarkAsComplete(evalContext *EvalContext, notifierState *NotifierState) error {
+	not := notifierState.notifier
```
Comment (bergquist) [inline pkg/services/alerting/notifier.go]: we dont need a new local var anymore.

File: pkg/services/alerting/notifier.go [hunk 6]
```diff
@@ -41,61 +40,82 @@ type notificationService struct {
...
+	cmd := &m.SetAlertNotificationStateToCompleteCommand{
+		State: notifierState.state,
+	}
 
-			if evalContext.IsTestRun {
-				return nil
-			}
+	if err = bus.DispatchCtx(evalContext.Ctx, cmd); err != nil {
+		if err == m.ErrAlertNotificationStateVersionConflict {
+			n.log.Error("notification state out of sync", "id", not.GetNotifierId())
```
Comment (bergquist) [inline pkg/services/alerting/notifier.go]: improve error message.

File: pkg/services/alerting/notifier.go [hunk 7]
```diff
@@ -41,61 +40,82 @@ type notificationService struct {
...
-			n.log.Debug("trying to send notification", "id", not.GetNotifierId())
+	err := not.Notify(evalContext)
 
-			// Verify that we can send the notification again
-			// but this time within the same transaction.
-			if !evalContext.IsTestRun && !not.ShouldNotify(ctx, evalContext) {
-				return nil
-			}
+	if err != nil {
+		n.log.Error("failed to send notification", "id", not.GetNotifierId())
```
Comment (bergquist) [inline pkg/services/alerting/notifier.go]: include error in log statement.

File: pkg/services/alerting/notifiers/base.go [hunk 8]
```diff
@@ -46,54 +45,49 @@ func NewNotifierBase(model *models.AlertNotification) NotifierBase {
 	}
 }
 
-func defaultShouldNotify(context *alerting.EvalContext, sendReminder bool, frequency time.Duration, journals []models.AlertNotificationJournal) bool {
+func defaultShouldNotify(context *alerting.EvalContext, sendReminder bool, frequency time.Duration, notificationState *models.AlertNotificationState) bool {
```
Comment (bergquist) [inline pkg/services/alerting/notifiers/base.go]: move this code to `ShouldNotify` instead.

File: pkg/services/alerting/result_handler.go [hunk 9]
```diff
@@ -67,6 +67,11 @@ func (handler *DefaultResultHandler) Handle(evalContext *EvalContext) error {
 			}
 
 			handler.log.Error("Failed to save state", "error", err)
+		} else {
+
+			// StateChanges is used for de dupping alert notifications
+			// when two servers are raising.
+			evalContext.Rule.StateChanges = cmd.Result.StateChanges
```
Comment (bergquist) [inline pkg/services/alerting/result_handler.go]: write better comments for this.

File: pkg/services/sqlstore/alert_notification.go [hunk 10]
```diff
@@ -229,44 +237,127 @@ func UpdateAlertNotification(cmd *m.UpdateAlertNotificationCommand) error {
...
 	})
 }
 
-func CleanNotificationJournal(ctx context.Context, cmd *m.CleanNotificationJournalCommand) error {
-	return inTransactionCtx(ctx, func(sess *DBSession) error {
-		sql := "DELETE FROM alert_notification_journal WHERE alert_notification_journal.org_id = ? AND alert_notification_journal.alert_id = ? AND alert_notification_journal.notifier_id = ?"
-		_, err := sess.Exec(sql, cmd.OrgId, cmd.AlertId, cmd.NotifierId)
-		return err
+func GetAlertNotificationState(ctx context.Context, cmd *m.GetNotificationStateQuery) error {
+	return withDbSession(ctx, func(sess *DBSession) error {
```
Comment (bergquist) [inline pkg/services/sqlstore/alert_notification.go]: this should be within one transaction

File: pkg/services/sqlstore/migrations/alert_mig.go [hunk 11]
```diff
@@ -107,4 +107,31 @@ func addAlertMigrations(mg *Migrator) {
...
+			{Cols: []string{"org_id", "alert_id", "notifier_id"}, Type: UniqueIndex},
+		},
+	}
+
+	mg.AddMigration("create alert_notification_state table v1", NewAddTableMigration(alert_notification_state))
+	mg.AddMigration("add index alert_notification_state org_id & alert_id & notifier_id",
+		NewAddIndexMigration(alert_notification_state, alert_notification_state.Indices[0]))
+
+	mg.AddMigration("Add alert_rule_state_updated_version to alert_notification_state", NewAddColumnMigration(alert_notification_state, &Column{
+		Name: "alert_rule_state_updated_version", Type: DB_BigInt, Nullable: true,
```
Comment (bergquist) [inline pkg/services/sqlstore/migrations/alert_mig.go]: no need for extra migration. move this to initial table.

**Review by bergquist — COMMENTED** [review]
Make sure we test sending test notifications before merging this.

File: pkg/models/alert_notifications.go (hunk 1 above)
Comment (bergquist) [reply pkg/models/alert_notifications.go]: :heavy_check_mark:

File: pkg/models/alert_notifications.go (hunk 2 above)
Comment (bergquist) [reply pkg/models/alert_notifications.go]: :heavy_check_mark:

File: pkg/services/alerting/interfaces.go (hunk 4 above)
Comment (bergquist) [reply pkg/services/alerting/interfaces.go]: :heavy_check_mark:

File: pkg/services/alerting/notifier.go (hunk 5 above)
Comment (bergquist) [reply pkg/services/alerting/notifier.go]: :heavy_check_mark:

File: pkg/services/alerting/notifier.go (hunk 6 above)
Comment (bergquist) [reply pkg/services/alerting/notifier.go]: :heavy_check_mark:

File: pkg/services/alerting/notifier.go (hunk 7 above)
Comment (bergquist) [reply pkg/services/alerting/notifier.go]: :heavy_check_mark:

File: pkg/services/alerting/notifiers/base.go (hunk 8 above)
Comment (bergquist) [reply pkg/services/alerting/notifiers/base.go]: :heavy_check_mark:

File: pkg/services/sqlstore/migrations/alert_mig.go (hunk 11 above)
Comment (bergquist) [reply pkg/services/sqlstore/migrations/alert_mig.go]: :heavy_check_mark:

File: pkg/services/sqlstore/alert_notification.go (hunk 10 above)
Comment (marefr) [reply pkg/services/sqlstore/alert_notification.go]: :heavy_check_mark:

**Pull request merged**